wasi/wasi.go

```diff
@@ -503,6 +535,23 @@ func (a *api) fd_close(ctx *wasm.HostFunctionCallContext, fd uint32) (err Errno)
 	return ErrnoSuccess
 }
 
+// RandomGet implements API.RandomGet
+func (a *api) RandomGet(ctx *wasm.HostFunctionCallContext, buf uint32, bufLen uint32) (errno Errno) {
+	if !ctx.Memory.ValidateAddrRange(buf, uint64(bufLen)) {
+		return ErrnoInval
+	}
+
+	random_bytes := make([]byte, bufLen)
+	_, err := a.randSource.Read(random_bytes)
+	if err != nil {
+		return ErrnoInval
```

Comment:
ErronoInval is for invalid arguments right? Is this the correct error for here?

Reply:
I would recommend take a look at spec and also WASI-libc and see how other impl handle errors

Reply:
* Wasmer return ErrnoIO for this case
https://github.com/wasmerio/wasmer/blob/53c2c29bed1cf1fee8c5616632ee8c2bccc27880/lib/wasi/src/syscalls/mod.rs#L2541

* wasi-libc just cast error fron syscal, and it can return bunch of them
https://man7.org/linux/man-pages/man2/getrandom.2.html

* wasmtime also maps syscal error into Errno
https://github.com/bytecodealliance/wasmtime/blob/853a025613e012c6c29002ddcccfced67073a8d0/crates/wasi-common/src/snapshots/preview_1.rs#L111

Reply:
FYI for bad memory offsets, we discussed offline to use ErrnoFault. Main thing is to document what is used and be consistent even if no one else is :D

Later, people can argue validly with these choices and probably we need to raise an issue on the wasi spec as it is underspecified. Main thing now is to document what you do.

Reply:
@codefromthecrypt thanks for info.
So, for invalid `buf`, `bufLen`  return error is `ErrnoFault`. Am I getting right?
And how to deal with errors from random source ?

Reply:
I think ErrnoIO for the source is fine for now. we can change it easily later if we find a most standard code

Reply:
I found the issue that has been punted quite a while so far https://github.com/WebAssembly/WASI/issues/215

Reply:
Error codes stuff gets you confused even more in functions that a more compilated than `random_get`. Spec need to address it.

Reply:
totally agree. appreciate if you can make your case on the spec issue because it was 2 years already so somehow they aren't prioritizing. maybe if more scream they will 🤷